authorization/token/service/token_service.go

```diff
@@ -0,0 +1,249 @@
+		// If the privileges are unchanged, then reset the token status
+		if loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_STALE) && resourceExistsInToken {
+			// Query for all of the token's privileges
+			privileges, err := s.Repositories().TokenRepository().ListPrivileges(ctx, tokenID)
+			if err != nil {
+				return "", errors.NewInternalError(ctx, err)
+			}
+
+			// First we recalculate any stale privileges
+			for _, priv := range privileges {
+				if priv.Stale {
+					// Retrieve the up to date scopes for the resource
+					scopes, err := s.Services().PrivilegeCacheService().ScopesForResource(ctx, priv.IdentityID, priv.ResourceID)
+					if err != nil {
+						return "", errors.NewInternalError(ctx, err)
+					}
+
+					scopesChanged := false
+
+					// Compare the scopes to those contained in the current token
+					for _, tokenPermission := range *tokenClaims.Permissions {
+						// Find the corresponding resource set ID in the token's permissions claim
+						if *tokenPermission.ResourceSetID == priv.ResourceID {
+							if !s.scopesEquivalent(tokenPermission.Scopes, scopes) {
+								scopesChanged = true
+								break
+							}
+						}
+					}
+
+					// If the scopes haven't changed, and the specified resouce ID is already contained in the current
+					// token, then reset the token status to valid and return an empty string
+					if !scopesChanged {
+						loadedToken.Status = 0
+						err = s.Repositories().TokenRepository().Save(ctx, loadedToken)
+						if err != nil {
+							return "", errors.NewInternalError(ctx, err)
+						}
+
+						return "", nil
+					}
+				}
+			}
+		}
+	}
+
+	// If we've gotten this far, it means that either no existing token was found, or the token that was found
+	// has been marked with status STALE and its privileges have changed, in either case we must generate a new token
+	perms := []token.Permissions{}
+
+	// Populate the permissions
+	// TODO populate permissions array
+
+	// Populate the scopes for the requested resource
+	resourceScopes, err := s.Services().PrivilegeCacheService().ScopesForResource(ctx, identity.ID, resourceID)
+	perm := &token.Permissions{
+		ResourceSetID: &resourceID,
+		Scopes:        resourceScopes,
+	}
+
+	perms = append(perms, *perm)
+
+	generatedToken, err := manager.GenerateUnsignedRPTTokenForIdentity(ctx, tokenClaims, *identity, &perms)
+	if err != nil {
+		return "", errors.NewInternalError(ctx, err)
+	}
+
+	// We need to extract the new jti claim (the Token ID) from the token, in order to create a database record for it
+	claims := generatedToken.Claims.(token.TokenClaims)
+
+	// Convert the new Token ID to a UUID value
+	newTokenID, err := uuid.FromString(claims.Id)
+	if err != nil {
+		return "", errors.NewInternalError(ctx, err)
+	}
+
+	// Create a new Token record in the database
+	newTokenRecord := &tokenRepo.Token{
+		TokenID:    newTokenID,
+		Status:     0,
+		TokenType:  tokenPkg.TOKEN_TYPE_RPT,
+		ExpiryTime: time.Unix(claims.ExpiresAt, 0),
```

Comment:
Please add a comment here saying we aren't using the expiry field at the moment.